tools/libSwiftScan/libSwiftScan.cpp

```diff
@@ -777,15 +882,618 @@ swiftscan_cache_compute_key(swiftscan_cas_t cas, int argc, const char **argv,
+        return failure(Result.takeError());
+      return Result->getNumOutputs();
+    }
+  }
+
+  return failure(createUnsupportedSchemaError());
+}
+
+bool swiftscan_cache_compilation_has_output_kind(
+    swiftscan_cached_compilation_t compilation, swiftscan_output_kind_t kind,
+    swiftscan_string_ref_t *error) {
+  auto failure = [&](llvm::Error E) {
+    *error =
+        swift::c_string_utils::create_clone(toString(std::move(E)).c_str());
+    return false;
+  };
+
+  auto CC = *unwrap(compilation);
+  if (!CC.isLoaded())
+    return failure(createUnloadedCompilationError());
+
+  *error = swift::c_string_utils::create_null();
+  auto &CAS = CC.LoadedProxy->getCAS();
+  {
+    swift::cas::CompileJobResultSchema Schema(CAS);
+    if (Schema.isRootNode(*CC.LoadedProxy)) {
+      auto Result = Schema.load(CC.LoadedProxy->getRef());
+      if (!Result)
+        return failure(Result.takeError());
+      auto OutputKind = getFileTypeFromScanOutputKind(kind);
+      bool HasOutput = false;
+      if (auto Err = Result->forEachOutput(
+              [&](swift::cas::CompileJobCacheResult::Output Output) {
+                if (Output.Kind == OutputKind)
+                  HasOutput = true;
+                return llvm::Error::success();
+              }))
+        return failure(std::move(Err));
+      return HasOutput;
+    }
+  }
+  {
+    clang::cas::CompileJobResultSchema Schema(CAS);
+    if (Schema.isRootNode(*CC.LoadedProxy)) {
+      // Only those three output kinds can be emitted.
+      clang::cas::CompileJobCacheResult::OutputKind OutputKind;
+      if (kind == SWIFTSCAN_OUTPUT_TYPE_CLANG_MODULE)
+        OutputKind = clang::cas::CompileJobCacheResult::OutputKind::MainOutput;
+      else if (kind == SWIFTSCAN_OUTPUT_TYPE_DEPENDENCIES)
+        OutputKind =
+            clang::cas::CompileJobCacheResult::OutputKind::Dependencies;
+      else if (kind == SWIFTSCAN_OUTPUT_TYPE_CACHED_DIAGNOSTICS)
+        OutputKind = clang::cas::CompileJobCacheResult::OutputKind::
+            SerializedDiagnostics;
+      else
+        return false;
+
+      auto Result = Schema.load(CC.LoadedProxy->getRef());
+      if (!Result)
+        return failure(Result.takeError());
+
+      bool HasOutput = false;
+      if (auto Err = Result->forEachOutput(
+              [&](clang::cas::CompileJobCacheResult::Output Output) {
+                if (Output.Kind == OutputKind)
+                  HasOutput = true;
+                return llvm::Error::success();
+              }))
+        return failure(std::move(Err));
+      return HasOutput;
+    }
+  }
+
+  return failure(createUnsupportedSchemaError());
+}
+
+swiftscan_cache_replay_instance_t
+swiftscan_cache_replay_instance_create(int argc, const char **argv,
+                                       swiftscan_string_ref_t *error) {
+  auto *Instance = new SwiftScanReplayInstance();
+
+  // Drop "-frontend" if it is the first argument.
+  if (argc > 0 && llvm::StringRef(argv[0]) == "-fronkend") {
```

Comment:
```suggestion
  if (argc > 0 && llvm::StringRef(argv[0]) == "-frontend") {
```

Reply:
Does this actually break things, in practice?

Reply:
Mostly, not. Just for writing tests in `swift` will be really hard without hacks to drop `swift-frontend` and `-frontend` because you need to invoke swift with `%target-swift-frontend` and maybe I can do the hack in test tool instead of behind C APIs.

Reply:
Removed for now as I don't see this code being triggered.